### PR TITLE
feat: Add support to forward subscriptionGroupIds

### PR DESCRIPTION
**Pull request by mmustafa-tse**
## Summary
 - We currently support forwarding subscriptionGroupIds mapped via user attributes in our Braze S2S forwarder only, a customer mentioned that Braze support this feature in their Braze and requested this as a feature to our kits, refer to our [docs](https://docs.mparticle.com/integrations/braze/event/#subscription-groups) regarding subscription group ids

 ## Testing Plan
 - [X] Was this tested locally? If not, explain why.
 - E2E and unit tested

 ## Reference Issue (For mParticle employees only.  Ignore if you are an outside contributor)
 - Closes https://go.mparticle.com/work/SQDSDKS-7258

### CODE REVIEW

File: src/BrazeKit-dev.js [hunk 1]
```diff
@@ -571,6 +572,26 @@ var constructor = function () {
...
+                kitLogger('braze.getUser().removeFromSubscriptionGroup', subscriptionGroupId);
+
+	            braze.getUser().removeFromSubscriptionGroup(subscriptionGroupId);
+            } else {
+                kitLogger('braze.getUser().addToSubscriptionGroup', subscriptionGroupId);
+
+	            braze.getUser().addToSubscriptionGroup(subscriptionGroupId);
+            }
+        }
+    }
```
Comment (alexs-mparticle) [inline src/BrazeKit-dev.js]: I think the nested if/else statements makes this function a bit hard to read. 

Also `typeof value !== 'boolean'` reads better than `!(typeof value === 'boolean')`.

P.S. Nit: Set up your linter!

```suggestion
    function setSubscriptionGroups(key, value) {
        var subscriptionGroupId = parsedSubscriptionGroupMapping[key];

        if (typeof value !== 'boolean') {
            kitLogger("Can't call setSubscriptionGroups on forwarder " +
                name +
                ', setSubscriptionGroups must set this value to a boolean');
            return;
        }

        var action = value ? 'addToSubscriptionGroup' : 'removeFromSubscriptionGroup';
        kitLogger('braze.getUser().' + action, subscriptionGroupId);
        braze.getUser()[action](subscriptionGroupId);
    }
```

File: src/BrazeKit-dev.js [hunk 2]
```diff
@@ -589,21 +610,25 @@ var constructor = function () {
...
+                    sanitizedValue
+                );
 
-            braze
-                .getUser()
-                .setCustomUserAttribute(sanitizedKey, sanitizedValue);
+                braze
+                    .getUser()
+                    .setCustomUserAttribute(sanitizedKey, sanitizedValue);
+            }
```
Comment (alexs-mparticle) [inline src/BrazeKit-dev.js]: Same here. I would rewrite this whole function to inverse the guards and make it cleaner.

```javascript
    function setUserAttribute(key, value) {
        if (key in DefaultAttributeMethods) {
            return setDefaultAttribute(key, value);
        }

        if (parsedSubscriptionGroupMapping[key]) {
            setSubscriptionGroups(key, value);
            return;
        }

        var sanitizedKey = getSanitizedValueForBraze(key);
        var sanitizedValue = getSanitizedValueForBraze(value);

        if (value != null && sanitizedValue == null) {
            return 'Value did not pass validation for ' + key;
        }

        kitLogger(
            'braze.getUser().setCustomUserAttribute',
            sanitizedKey,
            sanitizedValue
        );

        braze
            .getUser()
            .setCustomUserAttribute(sanitizedKey, sanitizedValue);
    }

```

File: src/BrazeKit-dev.js [hunk 3]
```diff
@@ -965,6 +994,18 @@ var constructor = function () {
...
+        let subscriptionGroupIds = {}; 
+      	const decodedSetting = subscriptionGroupSetting.replace(/&quot;/g, '"');
+        const parsedSetting = JSON.parse(decodedSetting)
+        for (let subscriptionGroupMap of parsedSetting) {
+            const key = subscriptionGroupMap.map
+            const value = subscriptionGroupMap.value
+            subscriptionGroupIds[key] = value
+        }
+        return subscriptionGroupIds
+    }
```
Comment (alexs-mparticle) [inline src/BrazeKit-dev.js]: As much as I want to, we should avoid using `const` and `let` in kits since we don't have transpiling set up.

```suggestion
        var subscriptionGroupIds = {}; 
      	var decodedSetting = subscriptionGroupSetting.replace(/&quot;/g, '"');
        var parsedSetting = JSON.parse(decodedSetting)
        for (let subscriptionGroupMap of parsedSetting) {
            var key = subscriptionGroupMap.map
            var value = subscriptionGroupMap.value
            subscriptionGroupIds[key] = value
        }
        return subscriptionGroupIds
    }
```

File: test/tests.js [hunk 4]
```diff
@@ -1107,6 +1117,45 @@ describe('Braze Forwarder', function() {
...
+        // get the decoded mapped subscriptionGroup
+        const parsedSubscriptionGroupMapping = mParticle.forwarder.decodeSubscriptionGroupMappings(subscriptionGroupMapping);
+
+        // set attribute subscriptionGroupTest1 with boolean value true should call Braze's addToSubscriptionGroup since it's mapped
+        mParticle.forwarder.setUserAttribute('subscriptionGroupTest1', true);
+        const mappedSubscriptionGroupId1 = parsedSubscriptionGroupMapping['subscriptionGroupTest1'];
+        window.braze.getUser().subscriptionGroup[mappedSubscriptionGroupId1].should.equal(true);
+
+        // set attribute subscriptionGroupTest2 with boolean value false should call Braze's removeFromSubscriptionGroup since it's mapped
+        mParticle.forwarder.setUserAttribute('subscriptionGroupTest2', false);
```
Comment (alexs-mparticle) [inline test/tests.js:1177]: I think this should be a separate test. You should have one test for the "positive" case, a separate one for the "negative" case, and then additional tests for specific edge cases.

File: test/tests.js [hunk 5]
```diff
@@ -1107,6 +1117,45 @@ describe('Braze Forwarder', function() {
...
+        const mappedSubscriptionGroupId2 = parsedSubscriptionGroupMapping['subscriptionGroupTest2'];
+        window.braze.getUser().subscriptionGroup[mappedSubscriptionGroupId2].should.equal(false);
+
+        // should log error if mapped attribute value is not type boolean
+        mParticle.forwarder.logger = {
+            verbose: function(msg) {
+                mParticle.forwarder.msg = msg;
+            },
+        };
+        mParticle.forwarder.setUserAttribute('subscriptionGroupTest1', 'test');
```
Comment (alexs-mparticle) [inline test/tests.js]: I would consider making this a separate test case as well.

File: test/tests.js [hunk 6]
```diff
@@ -1107,6 +1117,45 @@ describe('Braze Forwarder', function() {
         (window.braze.getUser().yearOfBirth === null).should.equal(true);
     });
 
+    it('it should set subscription group for mapped attributes when value is boolean', function() {
```
Comment (alexs-mparticle) [inline test/tests.js]: I would consider making specific unit tests for the new functions you've introduced if possible. I see this as an integration test, which is super helpful, but having tests for specific functions allows us to spot failures in specific functions makes it much easier to troubleshoot and catch bugs.

**Review by alexs-mparticle — CHANGES_REQUESTED** [review]

**Review by rmi22186 — COMMENTED** [review]
agree with all of Alex's points.  otherwise lgtm

File: test/tests.js (hunk 6 above)
Comment (rmi22186) [inline test/tests.js]: no need for the line to start with `it`

```suggestion
    it('should set subscription group for mapped attributes when value is boolean', function() {
```

File: test/tests.js [hunk 7]
```diff
@@ -1107,6 +1117,45 @@ describe('Braze Forwarder', function() {
...
+                apiKey: '123456',
+                subscriptionGroupMapping: subscriptionGroupMapping,
+            },
+            reportService.cb,
+            true,
+            null
+        );
+
+        // get the decoded mapped subscriptionGroup
+        const parsedSubscriptionGroupMapping = mParticle.forwarder.decodeSubscriptionGroupMappings(subscriptionGroupMapping);
```
Comment (rmi22186) [inline test/tests.js]: this should be its own test to determine that decodeSubscriptionGroupMappings is working properly

**Comment by mmustafa-tse** [comment]
@rmi22186 @alexs-mparticle - thanks all for the review, just changed per your recommendations, let me know how it looks

File: test/tests.js [hunk 8]
```diff
@@ -1107,6 +1117,106 @@ describe('Braze Forwarder', function() {
...
+        // initialize Braze kit with subscriptionGroupMappings
+        mParticle.forwarder.init(
+            {
+                apiKey: '123456',
+                subscriptionGroupMapping: subscriptionGroupMapping,
+            },
+            reportService.cb,
+            true,
+            null
+        );
```
Comment (rmi22186) [inline test/tests.js]: pretty sure this can be removed since you are only testing the method and you don't need to do a full initialization fo the forwarder to do so
```suggestion
```

**Review by rmi22186 — CHANGES_REQUESTED** [review]
very minor nit

File: test/tests.js (hunk 8 above)
Comment (mmustafa-tse) [reply test/tests.js]: @rmi22186 - just removed it, thanks for the feedback!

**Review by rmi22186 — APPROVED** [review]

File: src/BrazeKit-dev.js [hunk 9]
```diff
@@ -965,6 +991,18 @@ var constructor = function () {
...
+    function decodeSubscriptionGroupMappings(subscriptionGroupSetting) {
+        var subscriptionGroupIds = {}; 
+      	var decodedSetting = subscriptionGroupSetting.replace(/&quot;/g, '"');
+        var parsedSetting = JSON.parse(decodedSetting);
+        for (let subscriptionGroupMap of parsedSetting) {
+            var key = subscriptionGroupMap.map;
+            var value = subscriptionGroupMap.value;
+            subscriptionGroupIds[key] = value;
+        }
+        return subscriptionGroupIds;
```
Comment (rmi22186) [inline src/BrazeKit-dev.js:1003]: @mmustafa-tse I realize, we should actually put a try/catch block here just to be extra safe

Comment (rmi22186) [inline src/BrazeKit-dev.js:1003]: @mmustafa-tse  I realize, we should actually put a try/catch block here just to be extra safe.
